### PR TITLE
Add arm and arm64 builds of clang-8 to production

### PR DESCRIPTION
**Pull request by mattface**
Use the production build and boot infrastructure to test clang-8 for
a subset of architectures (armv7 and arm64).

As arm64 is a little noisy, and armv7 will likely be extremely noisy,
use a limited recipient list to not annoy the linux-next list.
@broonie is that the expected behaviour? ^

As the arch-complete job can't currently deal with different email
recipients for different builds, use a different tree name until
such time as it can.

FYI, I've also moved a gce builder over to production ready for
this but can add more if needed.

### CODE REVIEW

**Comment by gctucker** [comment]
Duplicating the `linux-next` tree has the following consequences:
* the same branch will be monitored twice - thanks to the new "last commit" file trick that should work but it's still not ideal and may lead to the git server bouncing if too many requests are sent to it in a small period of time
* 2 identical tarballs will be created for each linux-next tag, one for the regular tree and one for "next-clang"
* clang build and boot results will be stored in a separate category in the database, making it harder to search and deal with the data as they will effectively be hidden from the actual linux-next data (next-clang is just not a tree)
* this will need to be undone once clang builds are considered "good enough", causing more disruption and making it harder to search for all the clang builds in the database
* clang-9 still needs to be built on staging (I guess?) in which case this PR doesn't actually solve the initial problem of having a special build job

In summary, it's polluting the kernelci.org backend database just because the `kernel-arch-complete.sh` script doesn't have the flexibility to cope with this scenario.

What is the point of enabling clang builds in production if it's too experimental for the results to be treated as actual production data?  Can we not just wait until clang builds are in a good enough shape and enable this properly as part of the linux-next tree builds?  Or, if the actual issue is the noise level in build reports, can we not improve them so that they're shorter and easier to read? (say, include just the number of warnings and failures and add a link to a full log with the details)

**Comment by mattface** [comment]
I categorically disagree with that standpoint, and I don't see a strong reason to stop this being rolled out until the support arrives to do it in a cleaner manner.
I don't like this trend towards everything having to be some vision of "perfect" before it's allowed on production.

The benefits of moving clang builds into production is that I can move builders currently reserved for just clang on staging over to production so they can benefit all the trees, and so that clang builds get access to all of the boot labs to greatly increase their boot coverage.

Undoing this change will take seconds now as it's just config, indeed I can quickly change the config in this PR if someone says that all the build results should be going to the same recipients as the main next tree.

We can't expect any new compiler to have perfect results, instead we can help it towards that goal. What happens if we add some cutting edge version of gcc? Do we demote that to staging? It's all data, whether it's perfect or not shouldn't determine where it can be produced.

**Comment by gctucker** [comment]
I'm not asking for perfection, but it's true that we seem to have different standards for what we consider to be acceptable code and design.

You haven't answered my question about clang-9: will we still be running that build on staging?  If so, we should leave a dedicated builder there.

The main problem I have with this change is that I fail to understand why anyone would want to do this at all.  So I find it very hard to justify making any kind of compromise when the benefits aren't clear.  Here's a sample of what's being built on staging at the moment (seems to be only arm64, so there's no data with arm):
https://staging.kernelci.org/build/next-clang/branch/master/kernel/next-20190306/

Regular email reports are there to notify people when something starts breaking.  With the amount of build warnings and failures with clang-8, I understand that the emails may not be very useful, hence the reluctance to send it to the linux-next mailing list.  But, why do we want to push this in kernelci.org production if it's not useful?  Anyone with a computer can build a kernel with clang-8.

Finally, I still think we wouldn't be having this discussion if the build email reports were not listing every single compiler warning directly inside them.

**Comment by mattface** [comment]
I did state the benefits: Mostly to greatly increase the boot results available for clang built kernels, and a smaller benefit is to move dedicated build hardware to the general pool. Clang should be a first class citizen of kernelci.

For clang-9 I just havent put it in this PR as it's wasnt being tested on staging either, and I expected pushback for just clang-8. I will add it here if that clears up the issue.

Yes we agree build reports verbosity is an issue, there's a long out of date PR to change that but realistically it's not going to be fixed in short order. Neither is who receives the emails per compiler.

My frustration comes from improvements with a few drawbacks being blocked while the ideal situation is built. I don't think it has to be the best way every time.

I'd welcome comment from @khilman or @broonie or we will end up at a stalemate.

**Comment by broonie** [comment]
I think there's a few considerations here that are getting wrapped up together.

One is that there's a question about if staging is for staging changes to the KernelCI code/service or staging for builds that might be added to KernelCI, I think it should be the former. Right now we're in the position where people are actively using a job that's running on staging which means that it's more difficult to test things on staging - people will notice breakage, staging just becomes more of another production. As @mattface says it does also mean that we have extra builders tied up in staging and we've got much fewer boards in the mix. 

Another is the readiness of the builds themselves. For arm64 we do have a large splat of errors right now for allmodconfig but that's just a single error manifesting in many places, there's a patch already on the list for it. Otherwise the build is in reasonably good shape (Arnd's change to turn off asan helped a lot there) and once the last build failure is addressed I'd actually go further than Matt and consider just enabling clang/arm64 in the standard -next jobs. For arm it probably needs a bit more work, I didn't actually check recently, but then the whole point of the separate jobs is to support people who are actively working on this (and we know they're actively using the output). However all this is a bit moot as the current pull request is for a separate clang specific build so it'll not bother anyone who isn't interested in clang.

There's also the concern with having separate job names for the same tree potentially creating some confusion in the database. I can actually see that, but right now it's a potential rather than actual issue AFAICT. I'm obviously less up to speed with the backend than either @gctucker or @mattface but is this not something that could be in the worst case dealt with by throwing away the data from that job if we come up with a better way of doing things? To me this one definitely feels like something that we could live with.

**Comment by khilman** [comment]
I agree with @gctucker.  I really don't like having a separate (but identical) kernel tree just for building next with clang in production, and having to look in a separate tree in the UI etc.  Basically, I still think this is a short-term hack, and is only useful to the very few people who know what next-clang is for.

If we want clang to be a first class citizen, then the linux-next results for both gcc and clang should be visible with the linux-next results, not the "magic" next-clang.

Both @gctucker and I expressed similar objections to this when it was proposed to staging, and we reluctantly agreed to put it in staging as a stop gap until we fix the build-config yaml stuff.  Now we're having the same conversation again for production.

Can we please just fix the build-config.yaml so clang can be a first class citizen?

**Comment by broonie** [comment]
I still think there's multiple issues being conflated here.

@mattface could we perhaps split out the arm64 configuration into a separate pull request that just adds it to the standard -next job? We can then have it ready to merge as soon as the fix for the one remaining build error is in. That seems completely uncontroversial and it's where we want to end up so let's just do it and get it out of the way so it's not a distraction.

For the other stuff (just arm32 in the present pull request) I think it's more than just build-config.yaml changes that are needed - in cases where the build is not yet there or where the compiler is not yet released (both of which have people interested in them from the kernel side so seem within the KernelCI wheelhouse) don't we also need some development in the web front end if we want them as part of the same job - for one reason or another we'll be expecting failures that we probably don't want to report in the mainstream reports. that'll extend to the web UI as well as the e-mail reports. We'll need some way to display those separately, I had thought that the use of a separate tree was intended to deal with that issue.

**Comment by gctucker** [comment]
Agreed about the `arm64` part, I see there is now #106 which looks fine (still to be tested though).  Thanks @mattface.

About the `arm` builds and the frontend part, the tree name is essentially the job name on the frontend.  So having a `next-clang` tree would result in a separate job, as can be see on staging.kernelci.org.  The frontend can already cope reasonably with different compilers within the same job, here's an example:
https://staging.kernelci.org/build/gtucker/branch/kernelci-local/kernel/kernelci-local-snapshot-030-1-ge3a520bed955/
The only issue I can see there is that you need to expand each build entry to find out the compiler version, so it should probably be added next to the defconfig and arch name in the collapsed list view.

The main builds page looks OK:
https://staging.kernelci.org/build/?gtucker

> For the other stuff (just arm32 in the present pull request) I think it's more than just build-config.yaml changes that are needed - in cases where the build is not yet there or where the compiler is not yet released [...] don't we also need some development in the web front end if we want them as part of the same job

@broonie I believe I understand your question but I'm not entirely sure.  Is your concern that adding the `arm` clang builds in the same way as the `arm64` ones will cause a lot of noise in the build results for linux-next?  If that's the case, then yes using a separate tree (job) name works around it but with some downsides as discussed before.  There could be a better implementation to separate results on the frontend and in emails based on the compiler version, now I'm not sure if that's worth the effort as ultimately we probably do want only one build email per kernel revision with everything.

About deleting data for a "fake" next-clang tree, yes that can be done but unless there's some issue with storage or database size I think it would be wiser to leave it in to avoid breaking dependencies by mistake.  There's no clean way to delete data through the backend API, and manually changing the production database sounds like something we should only do if absolutely necessary.  So bottom line, I'm still in favor of not having any "fake job" next-clang data in there :)

**Comment by gctucker** [comment]
One more thing: how about adding clang-8 builds to the main next job but only with multi_v7_defconfig and maybe allmodconfig to start with?

**Comment by gctucker** [comment]
@mattface Could you please close this PR and if needed create a new one to add `arm` in https://github.com/kernelci/kernelci-core as we're shutting down this kernelci-core-staging repo this week?

The `arm64` part is cleanly covered in the regular linux-next tree by #104 (which should also move to kernelci-core).

**Comment by broonie** [comment]
@gctucker I do think there is a good use case for being able to support reporting for some compilers separately to the main reports for use with testing bleeding edge compilers. This is something that some kernel people already look at manually, we want to make sure that if the compilers are going to do something that causes trouble we get it fixed one way or another before the compiler is released so it doesn't get in the way of day to day kernel work, and it's also useful for LTS releases as it maximizes the chance that the kernel is going to work with newer compilers years after release.

**Comment by mattface** [comment]
As discussed and agreed in the weekly meeting, I'll close this one and bring the armv7 builds back to the table once the backend can better handle specific email recipients

**Comment by gctucker** [comment]
@broonie @mattface I've created this issue about filtering compiler versions in build reports:
https://github.com/kernelci/kernelci-backend/issues/119
